File: Algorithms/205.同构字符串/isIsomorphic.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        return isIsomorphicHelper(s, t) && isIsomorphicHelper(t, s);
    }

    private boolean isIsomorphicHelper(String s, String t) {
        int[] map = new int[128];
        char[] cs = s.toCharArray(), ct = t.toCharArray();
        for (int i = 0; i < s.length(); ++i)
            if (map[cs[i]] == 0)
                map[cs[i]] = ct[i];
            else if (map[cs[i]] != ct[i])
                return false;
        return true;
    }

}

File: Algorithms/205/isIsomorphic.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        s2t_map = {}
        t2s_map = {}
        for i in range(len(s)):
            if s[i] not in s2t_map and t[i] not in t2s_map:
                s2t_map[s[i]] = t[i]
                t2s_map[t[i]] = s[i]
            elif s[i] in s2t_map and s2t_map[s[i]] != t[i] or t[i] in t2s_map and t2s_map[t[i]] != s[i]:
                return False
        return True
